### PR TITLE
[Consensus] `EventLoop` internal queues update

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/dapperlabs/flow-go/issues/6230

### Context

This PR implements changes to internal queues for `EventLoop`. **It also includes a very important fix and tests for `NewestQcTracker`.** 

To guarantee liveness we can't afford to naively drop messages when queue is full(as it is in any `Engine`). That's why we will use different mechanism to process messages without blocking caller for too long.

In `EventLoop` we use two different mechanisms to deliver events:
1. Buffered channels for proposals and partial TC events, due to small number of objects
2. Newest QC and TC trackers which drop messages with lower view that we have already seen, this is needed since number of such messages is large.

### CODE REVIEW

File: consensus/hotstuff/eventhandler/event_handler_test.go [hunk 1]
```diff
@@ -927,6 +927,89 @@ func (es *EventHandlerSuite) TestOnReceiveProposal_ProposalForActiveView() {
...
+	require.NoError(es.T(), err)
+
+	// partial TC shouldn't trigger view change
+	require.Equal(es.T(), es.initView, es.paceMaker.CurView(), "incorrect view change")
+	// we don't want to create timeout if partial TC was delivered for view different than active one.
+	es.communicator.AssertNotCalled(es.T(), "BroadcastTimeout", mock.Anything)
+}
+
+// TestOnPartialTcCreated_QcAndTcProcessing tests that EventHandler processes QC and TC included in hotstuff.PartialTcCreated
+// data structure.
```
Comment (jordanschalm) [inline consensus/hotstuff/eventhandler/event_handler_test.go:973]:
```suggestion
// TestOnPartialTcCreated_QcAndTcProcessing tests that EventHandler processes QC and TC included in hotstuff.PartialTcCreated
// data structure. This tests cases like the following example:
// * the pacemaker is in view 10
// * we observe a partial timeout for view 11 with a QC for view 10
// * we should change to view 11 using the QC, then broadcast a timeout for view 11
```

File: consensus/hotstuff/eventhandler/event_handler_test.go [hunk 2]
```diff
@@ -927,6 +927,89 @@ func (es *EventHandlerSuite) TestOnReceiveProposal_ProposalForActiveView() {
...
+	lastCall := es.communicator.Calls[len(es.communicator.Calls)-1]
+	// the last call is BroadcastTimeout
+	require.Equal(es.T(), "BroadcastTimeout", lastCall.Method)
+	timeoutObject, ok := lastCall.Arguments[0].(*model.TimeoutObject)
+	require.True(es.T(), ok)
+	// it should broadcast a TO with same view as partialTc.View
+	require.Equal(es.T(), partialTc.View, timeoutObject.View)
+}
+
+// TestOnPartialTcCreated_NotActiveView tests that we don't create timeout object if partial TC was delivered for non-current view.
```
Comment (jordanschalm) [inline consensus/hotstuff/eventhandler/event_handler_test.go:956]:
```suggestion
// TestOnPartialTcCreated_NotActiveView tests that we don't create timeout object if partial TC was delivered for a past, non-current view.
// NOTE: it is not possible to receive a partial timeout for a FUTURE view, unless the partial timeout contains
// either a QC/TC allowing us to enter that view, therefore that case is not covered here. 
// See TestOnPartialTcCreated_QcAndTcProcessing instead.
```

File: consensus/hotstuff/eventloop/event_loop.go [hunk 3]
```diff
@@ -33,17 +38,27 @@ var _ component.Component = (*EventLoop)(nil)
...
+	// we will use a buffered channel to avoid blocking of caller
+	// we can't afford to drop messages since it undermines liveness, but we also want to avoid blocking of compliance
+	// engine. We assume that we should be able to process proposals faster than compliance engine feeds them, worst case
+	// we will fill the buffer and block compliance engine worker but that should happen only if compliance engine receives
+	// large number of blocks in short period of time(when catching up for instance).
+	proposals := make(chan *model.Proposal, 1000)
+	// we will use a buffered channel to avoid blocking of caller
+	// we will create at most one partial TC per view and only in recovery path
+	// since we don't expect a large number of messages we are using a buffered channel.
+	partialTcCreated := make(chan *hotstuff.PartialTcCreated, 10)
```
Comment (jordanschalm) [inline consensus/hotstuff/eventloop/event_loop.go:50]: I would like to add some visibility into these channels, now that they can store a significant backlog of messages. I'll add a note to the metrics issue.

File: consensus/hotstuff/eventloop/event_loop.go [hunk 4]
```diff
@@ -33,17 +38,27 @@ var _ component.Component = (*EventLoop)(nil)
...
 func NewEventLoop(log zerolog.Logger, metrics module.HotstuffMetrics, eventHandler hotstuff.EventHandler, startTime time.Time) (*EventLoop, error) {
-	proposals := make(chan *model.Proposal)
-	quorumCertificates := make(chan *flow.QuorumCertificate, 1)
-	timeoutCertificates := make(chan *flow.TimeoutCertificate, 1)
+	// we will use a buffered channel to avoid blocking of caller
+	// we can't afford to drop messages since it undermines liveness, but we also want to avoid blocking of compliance
+	// engine. We assume that we should be able to process proposals faster than compliance engine feeds them, worst case
+	// we will fill the buffer and block compliance engine worker but that should happen only if compliance engine receives
+	// large number of blocks in short period of time(when catching up for instance).
+	proposals := make(chan *model.Proposal, 1000)
```
Comment (jordanschalm) [inline consensus/hotstuff/eventloop/event_loop.go:46]:
```suggestion
	// TODO(active-pacemaker) add metrics for length of inbound channels
	proposals := make(chan *model.Proposal, 1000)
```

File: consensus/hotstuff/tracker/tracker_test.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+// TestNewestQCTracker_Track this test is needed to make sure that concurrent updates on NewestQCTracker are performed correctly,
+// and it always tracks the newest QC, especially in scenario of shared access. This test is structured in a way that it
+// starts multiple goroutines that will try to submit their QCs simultaneously to the tracker. Once all goroutines are started
+// we will use a wait group to execute all operations as concurrent as possible, after that we will observe if resulted value
+// is indeed expected. This test will run multiple times.
+func TestNewestQCTracker_Track(t *testing.T) {
+	tracker := NewNewestQCTracker()
+	samples := 20
+	times := 20
```
Comment (jordanschalm) [inline consensus/hotstuff/tracker/tracker_test.go:27]:
```suggestion
	samples := 20 // number of concurrent updates per test case
	times := 20   // number of times we run the test case
```

**Review by jordanschalm — APPROVED** [review]
Nice work

**Review by AlexHentschel — APPROVED** [review]
nice. Very clean. Great and thorough tests 👏 
Only a few stylistic comments.

**Comment by durkmurder** [comment]
bors merge

**Comment by durkmurder** [comment]
bors merge

**Comment by durkmurder** [comment]
bors cancel

**Comment by durkmurder** [comment]
bors merge

**Pull request merged**